### PR TITLE
Rough draft of an overhaul of the SDK

### PR DESCRIPTION
**Pull request by justinhopper**
Just looking for comment on the direction I'm taking the SDK, with it mostly focused on example programs of how to use the API, instead of a command line all-in-one tool. I can see the use for either approach, so let me know which direction is more useful. Afterward, I'd like to continue to ferret out the issues between the SDK and the API, as noted in the API notes file.

### CODE REVIEW

**Comment by justinhopper** [comment]
Couple of notes, that might help explain things quicker:
1) All of the machine specs calls should of course be moved to a proper function call, but right now that's all kinda shoehorned in, until we can move that functionality to the API server.
2) Most of the cluster and node calls are only set up for DigitalOcean, as that was the only key I had to use. I think the GCE one works, as I tested a build on that provider with my personal key once.
3) I'm working on building out the examples, along with proper testing, for AWS and the other providers. I just got the keys for AWS today.

**Comment by ntfrnzn** [comment]
I think the examples programs are a good idea, and a more likely pattern for users than the command-line utility.  Examples are great. If you could, please run `go fmt` over these files, and take a look at the `golint` suggestions.  Underscores look strange to me now in go programs.

File: example/cluster_create_digitalocean.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+func main() {
+    // Set up HTTP client with API token and URL
+    token := os.Getenv("SPC_API_TOKEN")
+    endpoint := os.Getenv("SPC_BASE_API_URL")
+    client := spio.NewClient(token, endpoint)
+
+    var ssh_keysetid int
+    fmt.Sscanf(os.Getenv("SPC_SSH_KEYSET"), "%d", &ssh_keysetid)
+    var do_keysetid int
+    fmt.Sscanf(os.Getenv("SPC_DO_KEYSET"), "%d", &do_keysetid)
```
Comment (ntfrnzn) [inline example/cluster_create_digitalocean.go]: Would be great to have an example of "how to look up the keyset id" and wrap that into something more user friendly.  The primary key integer where stackpointcloud stores the keyset is pretty meaningless to the end user.

File: example/cluster_create_digitalocean.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        fmt.Println(opt)
+    }
+
+    var node_size string
+    fmt.Printf("Enter node size: ")
+    fmt.Scanf("%s", &node_size)
+
+    // Validate machine type selection
+    found := false
+    for _, opt := range m_options {
```
Comment (ntfrnzn) [inline example/cluster_create_digitalocean.go]: can we wrap this into a function, "validateMachineType(provider, node_size string) error" returning nil if its ok?

File: example/cluster_create_digitalocean.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        if node_size == opt {
+            found = true
+            break
+        }
+    }
+    if found != true { 
+        fmt.Printf("Invalid option: %s\n", node_size)
+        os.Exit(1)
```
Comment (ntfrnzn) [inline example/cluster_create_digitalocean.go]: log.Fatalf is equivalent, and a little more standard.

File: example/cluster_create_gce.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    new_solution := spio.Solution{Solution: "helm_tiller"}
+    new_cluster := spio.Cluster{Name: cluster_name,
+                                Provider: provider,
+                                ProviderKey: gce_keysetid,
+                                MasterCount: 1,
+                                MasterSize: node_size,
+                                WorkerCount: 2,
+                                WorkerSize: node_size,
+                                Region: "us-west1-a",
+                                State: "draft",
```
Comment (ntfrnzn) [inline example/cluster_create_gce.go]: State: "draft" seems really unnecessary.  I'd call this requirement a server-side issue :(

File: example/cluster_create_gce.go [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+const orgid = 111
+const provider = "gce"
+const cluster_name = "Test GCE Cluster"
+
+func main() {
+    // Set up HTTP client with API token and URL
+    token := os.Getenv("SPC_API_TOKEN")
+    endpoint := os.Getenv("SPC_BASE_API_URL")
+    client := spio.NewClient(token, endpoint)
```
Comment (ntfrnzn) [inline example/cluster_create_gce.go]: want sensible error handling if these aren't defined.  The error reporting should probably go into `spio.NewClientFromEnvironment()` returning (*APIClient, error)

File: example/cluster_delete.go [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    if len(clusters) == 0 {
+        fmt.Printf("Sorry, no clusters defined yet\n")
+        os.Exit(0)
+    }
+
+    var clusterid int
+    fmt.Printf("Enter cluster ID to delete: ")
+    fmt.Scanf("%d", &clusterid)
+
+    client.DeleteCluster(orgid, clusterid)
```
Comment (ntfrnzn) [inline example/cluster_delete.go]: Checking the error is good practice.

File: example/cluster_create_digitalocean.go (hunk 1 above)
Comment (justinhopper) [reply example/cluster_create_digitalocean.go]: There is the keysets_list.go that can be run to find the different keysets loaded into the SPC API, and it was built for this purpose. But we definitely need some sort of documentation instructing the user to run that to find the keys, and probably to run the keysets_create_* scripts to load the keys (or load them manually in the web panel for the API)

File: example/cluster_create_digitalocean.go (hunk 2 above)
Comment (justinhopper) [reply example/cluster_create_digitalocean.go]: Yes, I think the whole machine type selection process should be wrapped into functions, one for fetching the list of valid machine types, and another to validate that choice. But, I wanted to wait and see what you guys wanted to do about a way to remotely call the machine types first.

File: example/cluster_create_digitalocean.go (hunk 3 above)
Comment (justinhopper) [reply example/cluster_create_digitalocean.go]: I wasn't using any of the log.* functions in anything, but looking now, it does what two lines of my code does in just one.  panic might be another option, and would work, but the word "panic" just seems so earth-shattering...

File: example/cluster_create_gce.go (hunk 4 above)
Comment (justinhopper) [reply example/cluster_create_gce.go]: Yes, it definitely is.  Waiting on this one too, waiting to see what we can do on the API side.

File: example/cluster_create_gce.go (hunk 5 above)
Comment (justinhopper) [reply example/cluster_create_gce.go]: I actually added this yesterday, but didn't send it up yet.  I ran into the real-life consequence: program panic when the shell environment wasn't set.  And yes, with all these scripts using the same call over and over, best to roll those into a function call that grabs what's needed from the environment, handles missing values, etc.  Will push this change shortly.

File: example/cluster_delete.go (hunk 6 above)
Comment (justinhopper) [reply example/cluster_delete.go]: I wasn't doing any error checking on the user input in any of the scripts, if that's what you meant? I wasn't sure how far we wanted to take this, how robust an example script should be when checking the user input. Give me some idea of where we'd want to be on that.

**Comment by justinhopper** [comment]
Closing out this old pull request, the new one is on branch "alpha"

**Pull request closed without merging**